### PR TITLE
Support Backward Compatibility with ElasticSearch 6.8 & 7.x versions

### PR DESCRIPTION
**Pull request by shwetathareja**
### Description
It is draft PR to discuss the approach for supporting backward compatibility with ElasticSearch nodes. 
The way version compatibility logic is implemented in 6.8 & 7.x, the OpenSearch has to return a 7.x version for it pass. OpenSearch node will return 7.10.2 if the remote node is ElasticSearch node otherwise 1.0.0

Note: Couple of tests are failed. I did basic sanity by launching mixed cluster using docker to ensure ElasticSearch & OpenSearch nodes are able to connect to each other. Raising it early to get feedback on the approach.
 
### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/669
 
### Check List
- [ ] Add logic to return BC ES version during handshake
- [ ] Remove masking logic
- [ ] Add logic to ensure OpenSearch is treated as successor to ES versions
- [ ] Besides handshake, JoinExecutor logic also expects compatible version in DiscoveryNode object otherwise node will fail to connect to master
- [ ] Add checks to fail handshake if ES node >= 7.10.3 (an actual ES node) as those version could be incompatible with OpenSearch
- [ ] Verify any other class which is calling writeVersion if it depends on localNode version
- [ ] All tests pass
- [ ] Run BWC tests
- [ ] Document the behavior of APIs when mixed nodes are present in the cluster
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by shwetathareja** [comment]
Approach during Handshake:

1. HandshakeRequest sends the minCompatibleVersion of the local node as it is not aware of the remote node version yet
2. During HandshakeResponse remote node sends its actual version, which is then set as the channelVersion.
3. The current proposal in the PR is to use 6.7.99 as the minCompatibleVersion for OpenSearch 1.x instead of 6.8 as this helps in (indirectly) identifying the remote node version during handleHandshake method itself and then send appropriate version (BC ES version/ OpenSearch 1.x version) in response. The advantage of doing is early identification of remote node version as otherwise if OpenSearch node also sends 6.8.0, there is no way to differentiate ES 7.x version from OpenSearch version and OpenSearch node will end up sending BC version to both ES & OpenSearch remote node. Sending only BC version to ElasticSearch node provide easy deprecation path for this BC version logic in OpenSearch 2.0.0 (which can expect upgrade from only Opensearch 1.x, though this is not finalized yet). Also 6.7 is already EOL, so there is no way there will be a real 6.7.99 version from ElasticSearch.

Callout:
1. Also considered sending a pseudo version something like 7.10.99 in response, this could help if customer want to identify OpenSearch nodes in APIs like _cat/nodes, Thoughts?
2.  The response of some API like _cat/nodes will show BC if the request is routed via ElasticSearch node. Pending list of all the APIs impacted.

**Review by dblock — COMMENTED** [review]
This looks good to me!

Note that I don't understand a bunch of things this old code was doing so don't take my LGTM as anything of value :) For my own education, could you please explain what we're losing by deleting this whole MASK business?

**Comment by shwetathareja** [comment]
@dblock : we added this masking logic earlier so that OpenSearch version id is higher than ES version ids but that is not working out as ES minCompatibility logic doesn't allow any higher version but only one higher version is accepted and as we need to support 6.8 as well, so BC version from OpenSearch has to be 7.x

File: server/src/main/java/org/opensearch/Version.java [hunk 1]
```diff
@@ -82,18 +85,25 @@ public static Version fromId(int id) {
         if (known != null) {
             return known;
         }
-        return fromIdSlow(id);
+        return fromIdSlow(id, false);
     }
 
-    private static Version fromIdSlow(int id) {
+    public static Version fromLegacyId(int id) {
```
Comment (dblock) [inline server/src/main/java/org/opensearch/Version.java:91]: I think we should call this `fromLegacyESId` to match `LegacyESVersion` and to be specific. One day there might be more legacy.

**Comment by dblock** [comment]
@shwetathareja This can be closed, no?

**Pull request closed without merging**